### PR TITLE
SI-5954 Adds implementation restriction preventing companions in package...

### PR DESCRIPTION
**Pull request by JamesIry**
... objects

Companion objects (and thus also case classes) in package objects caused an
assert about an overloaded symbol when everything was compiled twice. It's a
hairy problem that doesn't fit in 2.10.1. So this fix adds an implementation
restriction. It also has a test to make sure the error messages are clean
and reasonably friendly, and the test includes a commented out test in case
somebody thinks they've solved the underlying problem.

@review gkossakowski

### CODE REVIEW

**Comment by retronym** [comment]
Wouldn't the test be more straightforward as `neg` test?

**Comment by JamesIry** [comment]
@retronym Yup, didn't realize we had such things.

**Comment by JamesIry** [comment]
Re-submitted as a single commit, with simpler test in neg.

**Comment by JamesIry** [comment]
Oh, hey, the kitteh woke up and puked on my pr. (sigh)

**Comment by JamesIry** [comment]
PLS REBUILD ALL

**Comment by JamesIry** [comment]
> _< Builds clean on my box, but the kitteh don't lie, adding a value class to the test trips it up.

**Comment by JamesIry** [comment]
PLS REBUILD ALL

**Comment by JamesIry** [comment]
Looking at the tests that are failing, some are tests that were for bugs that duplicate this one except that the tests fail to actually test the problem. But other tests are for other tangential issues. They just happen to also fall afoul of the re-compile problem that this restriction is preventing.

So my proposal is to add a compiler switch that disables the restriction being added in this PR and modify some of the failing tests to use that flag. That way we don't lose valuable tests that exercise functionality that will one-day actually be useful when the re-compile problem is solved.

How say you @retronym, @paulp, @adriaanm, @gkossakowski ?

**Comment by retronym** [comment]
Sounds like a reasonable stepping stone to this musketeer, so long as we leave a ticket open.

**Comment by JamesIry** [comment]
PLS REBUILD ALL

File: test/files/neg/t5954.check [hunk 1]
```diff
@@ -0,0 +1,16 @@
+t5954.scala:36: error: implementation restriction: package object A cannot contain case class D. class D should be placed directly in package A instead.
```
Comment (jozic) [inline test/files/neg/t5954.check]: Shouldn't ${m} after .(dot) be capitalized?

```
  ...package object A cannot contain case class D. Class D should ...
```

**Comment by JamesIry** [comment]
Hmm, this desperately needs squashing.

**Comment by JamesIry** [comment]
@retronym lets get this puppy put to bed

**Comment by retronym** [comment]
:dog: Lucky! Go To Mattress!

**Comment by adriaanm** [comment]
:squirrel:!  (clarification: http://www.youtube.com/watch?v=bBWrMQVsuak)

**Comment by gkossakowski** [comment]
@JamesIry Apparently I was too slow to provide useful feedback. I'm not a big fan of introducing more compiler options but your balance seems to be around 0 so far so this is fine.

Also, thanks for following style of other implementation restrictions and for providing very good comments in the code. This is an example of how we should roll in scala/scala.

Also, https://issues.scala-lang.org/browse/SI-6109 is still open if you have apetite for something bigger :-)

**Comment by JamesIry** [comment]
I'm leaving the associated bug open because this implementation restriction is pretty arbitrary from a user point of view and it would be nice to solve the underlying problem.

**Pull request merged**